operator/scheduler/experiment.go

```diff
@@ -102,79 +102,78 @@ func (s *SchedulerClient) SubscribeExperimentEvents(ctx context.Context, grpcCli
 			logger.Info("Received experiment event with no k8s metadata so ignoring", "Experiment", event.ExperimentName)
 			continue
 		}
-		experiment := &v1alpha1.Experiment{}
-		err = s.Get(ctx, client.ObjectKey{Name: event.ExperimentName, Namespace: event.KubernetesMeta.Namespace}, experiment)
-		if err != nil {
-			logger.Error(err, "Failed to get experiment", "name", event.ExperimentName, "namespace", event.KubernetesMeta.Namespace)
-			continue
-		}
 
-		if !experiment.ObjectMeta.DeletionTimestamp.IsZero() {
-			logger.Info("Experiment is pending deletion", "experiment", experiment.Name)
-			if !event.Active {
-				retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-					latestExperiment := &v1alpha1.Experiment{}
-					err = s.Get(ctx, client.ObjectKey{Name: event.ExperimentName, Namespace: event.KubernetesMeta.Namespace}, latestExperiment)
-					if err != nil {
+		if !event.Active {
+			retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				ctxWithTimeout, cancel := context.WithTimeout(ctx, constants.K8sAPICallTimeout)
+				defer cancel()
+
+				latestExperiment := &v1alpha1.Experiment{}
+				err = s.Get(ctxWithTimeout, client.ObjectKey{Name: event.ExperimentName, Namespace: event.KubernetesMeta.Namespace}, latestExperiment)
+				if err != nil {
+					return err
+				}
+				if !latestExperiment.ObjectMeta.DeletionTimestamp.IsZero() { // Experiment is being deleted
+					// remove finalizer now we have completed successfully
+					latestExperiment.ObjectMeta.Finalizers = utils.RemoveStr(latestExperiment.ObjectMeta.Finalizers, constants.ExperimentFinalizerName)
+					if err := s.Update(ctxWithTimeout, latestExperiment); err != nil {
+						logger.Error(err, "Failed to remove finalizer", "experiment", latestExperiment.GetName())
 						return err
 					}
-					if !latestExperiment.ObjectMeta.DeletionTimestamp.IsZero() { // Experiment is being deleted
-						// remove finalizer now we have completed successfully
-						latestExperiment.ObjectMeta.Finalizers = utils.RemoveStr(latestExperiment.ObjectMeta.Finalizers, constants.ExperimentFinalizerName)
-						if err := s.Update(ctx, latestExperiment); err != nil {
-							logger.Error(err, "Failed to remove finalizer", "experiment", latestExperiment.GetName())
-							return err
-						}
-					}
-					return nil
-				})
-				if retryErr != nil {
-					logger.Error(err, "Failed to remove finalizer after retries")
 				}
+				return nil
+			})
+			if retryErr != nil {
+				logger.Error(err, "Failed to remove finalizer after retries")
```

Comment:
Can we have `!event.Active` but `latestExperiment.ObjectMeta.DeletionTimestamp.IsZero()` ? Because in that case we don't even attempt to remove the finalizer, and this error message might be confusing. For example, I don't know if RetryOnConflict doesn't return an error on context timeout. I suppose we still have the underlying error wrapped here.

Reply:
Also should we `continue` after logging here (rather than continuing with the status update)?

Reply:
actually there is a subtle behaviour here as `event.Active==False` could mean that some models that are part of an experiments are not ready anymore and therefore the whole experiment is not active. This overlaps with also deleting an experiment where `latestExperiment.ObjectMeta.DeletionTimestamp.IsZero()!=0`.

Actually we could `continue` if we managed to successfully remove the finalizer as the CR is going to be removed from k8s.

Reply:
> Also should we `continue` after logging here (rather than continuing with the status update)?

we need to detect first whether we removed the finalizer. I will leave it as a follow up PR though.